feat!: migrate to direct agentic API (v2) (#334)
markupai/content-guardian-action: CLAUDE.md

> # CLAUDE.md

Author: **Markup AI** detected issues:
- **Clarity / Flesch-Kincaid grade level (Severity: Medium)**: `CLAUDE.md`
*Explanation: Flesch-Kincaid grade level: 7. The Flesch-Kincaid Grade Level is a classic readability metric. You can think of it as the number of years of education required to understand a text.*

This file provides guidance to Claude Code (claude.ai/code) when working with code in this repository.

> ## What this project is

Author: **Markup AI** detected issues:
- **Consistency / Is your title case consistent? (Severity: High)**: `this`
*Explanation: Could you use the AP Stylebook for title case? Capitalize most words in titles.*
```suggestion
## What This project is
```
- **Consistency / Is your title case consistent? (Severity: High)**: `project`
*Explanation: Could you use the AP Stylebook for title case? Capitalize most words in titles.*
```suggestion
## What this Project is
```
- **Consistency / Is your title case consistent? (Severity: High)**: `is`
*Explanation: Could you use the AP Stylebook for title case? Capitalize most words in titles.*
```suggestion
## What this project Is
```

Runtime: Node 24 (see `.node-version`, `engines.node`). The published action runs `dist/index.js` (see `action.yml`).

## Commands

```bash
npm install              # install deps
npm test                 # run all vitest tests (live-smoke is auto-skipped)
npm run test:coverage    # tests + v8 coverage (used in CI; feeds SonarQube)
npm run lint:check       # eslint (strict-type-checked config)
npm run lint:fix
npm run format:check     # prettier
npm run format:fix
npm run type-check       # tsc --noEmit against tsconfig.json
npm run package          # rollup bundle → dist/index.js
npm run bundle           # format:fix + package (this is what CI's "check-dist" runs)
npm run all              # lint + coverage + bundle (full local CI gate)
npm run local-action     # run main.ts locally via @github/local-action using .env
```

Run a single test file: `npx vitest run test/path/to/file.test.ts`. Watch mode: `npx vitest`.

> Exercise the real API end-to-end (skipped without the key):

Author: **Markup AI** detected issues:
- **Tone / Could you replace this overused phrase? (Severity: High)**: `end`
*Explanation: Could you replace this overused phrase? Don’t use clichés as they can be hard to understand and might have lost their meaning. Replace them with clearer words and phrases.*
- **Clarity / Did you define the acronym in your content? (Severity: Medium)**: `API`
*Explanation: Did you define the acronym in your content? &#34;API&#34; (5)*

```bash
MARKUP_AI_LIVE_KEY=mat_... npx vitest run test/live-smoke.test.ts
```

> ## Critical: the `dist/` directory is committed

Author: **Markup AI** detected issues:
- **Consistency / Is your title case consistent? (Severity: High)**: `the`
*Explanation: Could you use the AP Stylebook for title case? Capitalize most words in titles.*
```suggestion
## Critical: The `dist/` directory is committed
```
- **Consistency / Is your title case consistent? (Severity: High)**: `directory`
*Explanation: Could you use the AP Stylebook for title case? Capitalize most words in titles.*
```suggestion
## Critical: the `dist/` Directory is committed
```
- **Consistency / Is your title case consistent? (Severity: High)**: `is`
*Explanation: Could you use the AP Stylebook for title case? Capitalize most words in titles.*
```suggestion
## Critical: the `dist/` directory Is committed
```
- **Consistency / Is your title case consistent? (Severity: High)**: `committed`
*Explanation: Could you use the AP Stylebook for title case? Capitalize most words in titles.*
```suggestion
## Critical: the `dist/` directory is Committed
```

> Because the action is consumed directly from this repo (`uses: markupai/content-guardian-action@v2` loads `dist/index.js`), the bundled output is checked in. The `check-dist.yml` workflow fails PRs whose `dist/` doesn't match a fresh `npm run bundle`. **Any source change must be followed by `npm run bundle` and a commit of the resulting `dist/` diff.** Don't hand-edit files under `dist/` — they are generated.

Author: **Markup AI** detected issues:
- **Tone / Did you use contractions? (Severity: High)**: `they`
*Explanation: Could you use contractions to make it less formal? Use contractions instead of the full form to make your content more conversational.*
```suggestion
Because the action is consumed directly from this repo (`uses: markupai/content-guardian-action@v2` loads `dist/index.js`), the bundled output is checked in. The `check-dist.yml` workflow fails PRs whose `dist/` doesn't match a fresh `npm run bundle`. **Any source change must be followed by `npm run bundle` and a commit of the resulting `dist/` diff.** Don't hand-edit files under `dist/` — they’re are generated.
```

## Architecture

> Entry chain: `dist/index.js` (rollup output) ← `src/index.ts` (calls `run()`) ← `src/main.ts` ← `src/action-runner.ts::runAction()`. All real orchestration lives in `action-runner.ts`.

Author: **Markup AI** detected issues:
- **Spelling and grammar / Could you capitalize the first word of the sentence? (Severity: High)**: `rollup`
*Explanation: Could you capitalize the first word of the sentence? In general, capitalize the first letter of a word that begins a sentence.*
```suggestion
Entry chain: `dist/index.js` (Rollup output) ← `src/index.ts` (calls `run()`) ← `src/main.ts` ← `src/action-runner.ts::runAction()`. All real orchestration lives in `action-runner.ts`.
```

The runner is a linear pipeline:

> 1. **Config** (`src/config/action-config.ts`) reads inputs/env via `@actions/core`, validates, and produces an `ActionConfig` with one required style input: `target`. Inputs fall back to env vars (e.g. `markup_ai_api_key` → `MARKUP_AI_API_KEY`).

Author: **Markup AI** detected issues:
- **Consistency / Is your comma usage consistent? (Severity: High)**: `,`
*Explanation: Could you remove the comma before &#34;and&#34; or &#34;or&#34;? Remove the comma before the conjunction unless one of the elements of the series contains a conjunction.*
```suggestion
1. **Config** (`src/config/action-config.ts`) reads inputs/env via `@actions/core`, validates and and produces an `ActionConfig` with one required style input: `target`. Inputs fall back to env vars (e.g. `markup_ai_api_key` → `MARKUP_AI_API_KEY`).
```
- **Clarity / Could you avoid Latin expressions? (Severity: High)**: `e.g.`
*Explanation: Could you avoid Latin expressions? Careful with Latin expressions. You might know exactly what they mean, but not everybody does.*
```suggestion
1. **Config** (`src/config/action-config.ts`) reads inputs/env via `@actions/core`, validates, and produces an `ActionConfig` with one required style input: `target`. Inputs fall back to env vars (for example, `markup_ai_api_key` → `MARKUP_AI_API_KEY`).
```
- **Spelling and grammar / Is this word spelled correctly? (Severity: High)**: `env`
*Explanation: Spelling: 'env' → 'end'. Is this word spelled correctly?*
```suggestion
1. **Config** (`src/config/action-config.ts`) reads inputs/env via `@actions/core`, validates, and produces an `ActionConfig` with one required style input: `target`. Inputs fall back to end vars (e.g. `markup_ai_api_key` → `MARKUP_AI_API_KEY`).
```
- **Spelling and grammar / Is this word spelled correctly? (Severity: High)**: `vars`
*Explanation: Spelling: 'vars' → 'cars'. Is this word spelled correctly?*
```suggestion
1. **Config** (`src/config/action-config.ts`) reads inputs/env via `@actions/core`, validates, and produces an `ActionConfig` with one required style input: `target`. Inputs fall back to env cars (e.g. `markup_ai_api_key` → `MARKUP_AI_API_KEY`).
```

> 2. **Bootstrap** — call `GET /style-agent/config` to read `style_agent_numeric_scoring` and assert style agent is enabled, then call `GET /style-agent/targets` and resolve the user's `target` input (by id or case-insensitive display_name) via `services/target-resolver.ts`. The resulting `AnalysisOptions` carries `{ targetId, targetDisplayName, numericScoringEnabled }`.

Author: **Markup AI** detected issues:
- **Scannability / Could you keep each list item as short as possible? (Severity: Medium)**: `Bootstrap`
*Explanation: Could you keep each list item as short as possible? Use 25 words or fewer. This list item has 27 words.*

3. **File discovery** (`src/strategies/file-discovery-strategies.ts`) — strategy pattern keyed on `github.context.eventName`:

>    - `push` → files touched in that commit (excluding deletions)

Author: **Markup AI** detected issues:
- **Consistency / Is the punctuation at the end of list elements consistent? (Severity: Medium)**: `files`
*Explanation: Is the punctuation in list elements consistent? This punctuation doesn’t clearly mark the end of a sentence.*

>    - `pull_request` → files changed in the PR

Author: **Markup AI** detected issues:
- **Consistency / Is the punctuation at the end of list elements consistent? (Severity: Medium)**: `    `
*Explanation: Is the punctuation in list elements consistent? There’s no punctuation at the end of this item.*

>    - `workflow_dispatch` / `schedule` → all repo files at `ref`

Author: **Markup AI** detected issues:
- **Consistency / Could you review the spacing around slashes? (Severity: Medium)**: `    `
*Explanation: Could you review the spacing around slashes? You’ve used spaces here.*
```suggestion
 /  - `workflow_dispatch` / `schedule` → all repo files at `ref`
```
- **Consistency / Is the punctuation at the end of list elements consistent? (Severity: Medium)**: `    `
*Explanation: Is the punctuation in list elements consistent? There’s no punctuation at the end of this item.*

>    - Unknown event types fall back to the push strategy with a warning.

Author: **Markup AI** detected issues:
- **Consistency / Is the punctuation at the end of list elements consistent? (Severity: Medium)**: `Un`
*Explanation: Is the punctuation in list elements consistent? There’s punctuation at the end of this item. Could you remove the punctuation mark?*

4. **Filter** to supported extensions (`SUPPORTED_EXTENSIONS` in `src/constants/index.ts`): `.dita .htm .html .markdown .md .mdown .mkd .text .txt .xml`.

> 5. **Analysis** (`src/services/api-service.ts`) — per file: read content, `POST /agents/ag_vYCPHsSQnnJj/run?wait=false` with `{ text, document_name, document_ref, target_id }`, then poll `GET /agents/workflows/{id}` every 2s (5-minute timeout). Files are processed with a concurrency cap of 5 (`MAX_CONCURRENT_FILES`). The style agent ID is hardcoded — it's a stable platform identifier, not customer-scoped.

Author: **Markup AI** detected issues:
- **Consistency / Could you review the spacing between numerals and units? (Severity: Medium)**: `2`
*Explanation: Could you review the spacing between numerals and units? You’ve left out spaces here.*
```suggestion
5. **Analysis** (`src/services/api-service.ts`) — per file: read content, `POST /agents/ag_vYCPHsSQnnJj/run?wait=false` with `{ text, document_name, document_ref, target_id }`, then poll `GET /agents/workflows/{id}` every 2 ss (5-minute timeout). Files are processed with a concurrency cap of 5 (`MAX_CONCURRENT_FILES`). The style agent ID is hardcoded — it's a stable platform identifier, not customer-scoped.
```
- **Scannability / Could you keep each list item as short as possible? (Severity: Medium)**: `Analysis`
*Explanation: Could you keep each list item as short as possible? Use 25 words or fewer. This list item has 35 words.*

6. **Post-analysis** (`src/services/post-analysis-service.ts`) — dispatch on event type:
   - `push` → `updateCommitStatus` (gated by `addCommitStatus`).
   - `pull_request` → `createOrUpdatePRComment` always; `createPRReviewComments` only if `addReviewComments`.

>    - `workflow_dispatch` / `schedule` → `createJobSummary` (writes `$GITHUB_STEP_SUMMARY` markdown).

Author: **Markup AI** detected issues:
- **Consistency / Could you review the spacing around slashes? (Severity: Medium)**: `    `
*Explanation: Could you review the spacing around slashes? You’ve used spaces here.*
```suggestion
 /  - `workflow_dispatch` / `schedule` → `createJobSummary` (writes `$GITHUB_STEP_SUMMARY` markdown).
```

7. **Outputs** — `event-type`, `files-analyzed`, `results` (JSON of `AnalysisResult[]`).
8. **Strict mode** — if `strict_mode: true` and any file failed to analyze, the action calls `core.setFailed("Some files were not analyzed.")` _after_ all post-analysis work has already run.

> ## Risk vs. numeric scoring

Author: **Markup AI** detected issues:
- **Consistency / Is your title case consistent? (Severity: High)**: `numeric`
*Explanation: Could you use the AP Stylebook for title case? Capitalize most words in titles.*
```suggestion
## Risk vs. Numeric scoring
```
- **Consistency / Is your title case consistent? (Severity: High)**: `scoring`
*Explanation: Could you use the AP Stylebook for title case? Capitalize most words in titles.*
```suggestion
## Risk vs. numeric Scoring
```

A single boolean — `numericScoringEnabled` from `GET /style-agent/config` — drives every output path.

- **Numeric on** (`style_agent_numeric_scoring: true`): result tables and commit statuses include `quality.score` (0–100). The summary averages the score across files.

> - **Numeric off** (default for most orgs): the action shows only severity counts (`H:_ M:_ L:_`) and an overall risk label derived from `issues[].severity` (`high` if any high present, else `medium`, else `low`, else `none`). No fabricated scores.

Author: **Markup AI** detected issues:
- **Scannability / Could you keep each list item as short as possible? (Severity: Medium)**: `Numeric`
*Explanation: Could you keep each list item as short as possible? Use 25 words or fewer. This list item has 29 words.*

> In both modes the `outputs.results` JSON contains the full per-file `AnalysisResult` (including raw issues), so downstream steps don't lose information.

Author: **Markup AI** detected issues:
- **Clarity / Could you add a comma after the introductory phrase? (Severity: High)**: `In`
*Explanation: Could you add a comma after the introductory phrase? If you use a comma after a prepositional or introductory phrase, your content will be easier to read.*

> ## API client (`src/services/markup-api-client.ts`)

Author: **Markup AI** detected issues:
- **Consistency / Is your title case consistent? (Severity: High)**: `client`
*Explanation: Could you use the AP Stylebook for title case? Capitalize most words in titles.*
```suggestion
## API Client (`src/services/markup-api-client.ts`)
```

> Thin fetch-based wrapper. Single `request<T>()` helper sets `Authorization: Bearer <key>` and `x-integration-id: markupai-content-guardian-action`, JSON-decodes the body, and throws `MarkupApiError(status, requestId, body)` on non-2xx. Retries once on 5xx and on raw network errors; never retries on 4xx. `isFatalApiError()` returns true for 401 / 403 / 5xx — these abort the whole run from `api-service.analyzeFile`, while non-fatal failures (timeouts, individual workflow `failed` status) just drop that file from the result list.

Author: **Markup AI** detected issues:
- **Spelling and grammar / Could you capitalize the first word of the sentence? (Severity: High)**: `returns`
*Explanation: Could you capitalize the first word of the sentence? In general, capitalize the first letter of a word that begins a sentence.*
```suggestion
Thin fetch-based wrapper. Single `request<T>()` helper sets `Authorization: Bearer <key>` and `x-integration-id: markupai-content-guardian-action`, JSON-decodes the body, and throws `MarkupApiError(status, requestId, body)` on non-2xx. Retries once on 5xx and on raw network errors; never retries on 4xx. `isFatalApiError()` Returns true for 401 / 403 / 5xx — these abort the whole run from `api-service.analyzeFile`, while non-fatal failures (timeouts, individual workflow `failed` status) just drop that file from the result list.
```

> ## Services and where to extend

Author: **Markup AI** detected issues:
- **Consistency / Is your title case consistent? (Severity: High)**: `where`
*Explanation: Could you use the AP Stylebook for title case? Capitalize most words in titles.*
```suggestion
## Services and Where to extend
```
- **Consistency / Is your title case consistent? (Severity: High)**: `to`
*Explanation: Could you use the AP Stylebook for title case? Capitalize most words in titles.*
```suggestion
## Services and where To extend
```
- **Consistency / Is your title case consistent? (Severity: High)**: `extend`
*Explanation: Could you use the AP Stylebook for title case? Capitalize most words in titles.*
```suggestion
## Services and where to Extend
```

> - `src/services/github-service.ts` — thin octokit wrapper (commit fetch, PR files, repo file walk, commit status with risk/numeric formatting).

Author: **Markup AI** detected issues:
- **Consistency / Could you start all list items with the same type of phrase? (Severity: Medium)**: `thin`
*Explanation: Could you start all list items with the same type of phrase? The list might be clearer if you start every item with the same phrase. &#34;thin&#34; (noun-phrase)*

> - `src/services/pr-comment-service.ts` — summary PR comment + per-issue inline review comments. Branches on `numericScoringEnabled` for table headers; uses `issue.position.text` / `issue.guideline_name` / `issue.suggestion(s)` from the agent response.

Author: **Markup AI** detected issues:
- **Consistency / Could you review the spacing around slashes? (Severity: Medium)**: `    `
*Explanation: Could you review the spacing around slashes? You’ve used spaces here.*
Suggestion: `  / /  `
- **Consistency / Could you start all list items with the same type of phrase? (Severity: Medium)**: `summary`
*Explanation: Could you start all list items with the same type of phrase? The list might be clearer if you start every item with the same phrase. &#34;summary&#34; (noun-phrase)*

> - `src/services/job-summary-service.ts` — `$GITHUB_STEP_SUMMARY` markdown for manual/scheduled runs.

Author: **Markup AI** detected issues:
- **Consistency / Could you start all list items with the same type of phrase? (Severity: Medium)**: `markdown`
*Explanation: Could you start all list items with the same type of phrase? The list might be clearer if you start every item with the same phrase. &#34;markdown&#34; (noun-phrase)*

> - `src/services/target-resolver.ts` — maps user input (id or display_name) to an enabled `StyleTarget`.

Author: **Markup AI** detected issues:
- **Consistency / Could you start all list items with the same type of phrase? (Severity: Medium)**: `maps`
*Explanation: Could you start all list items with the same type of phrase? The list might be clearer if you start every item with the same phrase. &#34;maps&#34; (noun-phrase)*

> - `src/utils/issue-utils.ts` — `computeIssueCounts`, `classifyRisk`, `aggregateCounts`, `aggregateRisk` and the `RISK_EMOJI` / `RISK_LABEL` maps.

Author: **Markup AI** detected issues:
- **Consistency / Could you review the spacing around slashes? (Severity: Medium)**: `    `
*Explanation: Could you review the spacing around slashes? You’ve used spaces here.*
Suggestion: `  /  `
- **Consistency / Could you start all list items with the same type of phrase? (Severity: Medium)**: `and`
*Explanation: Could you start all list items with the same type of phrase? The list might be clearer if you start every item with the same phrase. &#34;and&#34; (other)*

> - `src/utils/score-utils.ts` — numeric-mode helpers only (`getQualityStatus`, `calculateScoreSummary`).

Author: **Markup AI** detected issues:
- **Consistency / Could you start all list items with the same type of phrase? (Severity: Medium)**: `numeric`
*Explanation: Could you start all list items with the same type of phrase? The list might be clearer if you start every item with the same phrase. &#34;numeric-mode&#34; (noun-phrase)*

> - `src/constants/index.ts` is the single source for input names, env var names, event types, extensions, API URL, integration ID, and the hardcoded `STYLE_AGENT_ID`.

Author: **Markup AI** detected issues:
- **Clarity / Is this list of descriptions too long? (Severity: High)**: `the`
*Explanation: Is this list of descriptions too long? Too many descriptions can make a sentence difficult to read. Split up the sentence or create a list.*
- **Consistency / Is your comma usage consistent? (Severity: High)**: `,`
*Explanation: Could you remove the comma before &#34;and&#34; or &#34;or&#34;? Remove the comma before the conjunction unless one of the elements of the series contains a conjunction.*
```suggestion
- `src/constants/index.ts` is the single source for input names, env var names, event types, extensions, API URL, integration ID and and the hardcoded `STYLE_AGENT_ID`.
```
- **Spelling and grammar / Is this word spelled correctly? (Severity: High)**: `env`
*Explanation: Spelling: 'env' → 'end'. Is this word spelled correctly?*
```suggestion
- `src/constants/index.ts` is the single source for input names, end var names, event types, extensions, API URL, integration ID, and the hardcoded `STYLE_AGENT_ID`.
```
- **Clarity / Could this sentence be simpler? (Severity: Medium)**: `is`
*Explanation: Too complex? Your readers need content that is simple to read. Perhaps this sentence is too long?*
- **Clarity / Could this sentence be simpler? (Severity: Medium)**: `the`
*Explanation: Too complex? Your readers need content that is simple to read. Try to break up this list of descriptions.*
- _1 more issue(s) on this line_

> - `src/types/index.ts` defines our own types — `AgentRunRequest/Response`, `WorkflowStatus`, `OrganizationConfigResponse`, `StyleTarget`, `StyleScores`, `StyleAnalysis`, `AnalysisIssue`, `IssueCounts`, `RiskLevel`, `AnalysisResult`, `ActionConfig`, `AnalysisOptions`.

Author: **Markup AI** detected issues:
- **Consistency / Could you start all list items with the same type of phrase? (Severity: Medium)**: `defines`
*Explanation: Could you start all list items with the same type of phrase? The list might be clearer if you start every item with the same phrase. &#34;defines&#34; (verb-phrase)*

> ## Local development

Author: **Markup AI** detected issues:
- **Consistency / Is your title case consistent? (Severity: High)**: `development`
*Explanation: Could you use the AP Stylebook for title case? Capitalize most words in titles.*
```suggestion
## Local Development
```

> `npm run local-action` uses `@github/local-action` to run `src/main.ts` against `.env` (see `.env.example`). Env vars follow the GitHub Actions `INPUT_<NAME>` convention — and the runner is strict about it, so `INPUT_TARGET` keeps the underscore (the input name is `target`, not `style-guide` anymore). `.github/event.json` is a sample event payload for local PR runs.

Author: **Markup AI** detected issues:
- **Spelling and grammar / Could you capitalize the first word of the sentence? (Severity: High)**: `uses`
*Explanation: Could you capitalize the first word of the sentence? In general, capitalize the first letter of a word that begins a sentence.*
```suggestion
`npm run local-action` Uses `@github/local-action` to run `src/main.ts` against `.env` (see `.env.example`). Env vars follow the GitHub Actions `INPUT_<NAME>` convention — and the runner is strict about it, so `INPUT_TARGET` keeps the underscore (the input name is `target`, not `style-guide` anymore). `.github/event.json` is a sample event payload for local PR runs.
```
- **Spelling and grammar / Could you capitalize the first word of the sentence? (Severity: High)**: `is`
*Explanation: Could you capitalize the first word of the sentence? In general, capitalize the first letter of a word that begins a sentence.*
```suggestion
`npm run local-action` uses `@github/local-action` to run `src/main.ts` against `.env` (see `.env.example`). Env vars follow the GitHub Actions `INPUT_<NAME>` convention — and the runner is strict about it, so `INPUT_TARGET` keeps the underscore (the input name is `target`, not `style-guide` anymore). `.github/event.json` Is a sample event payload for local PR runs.
```
- **Spelling and grammar / Is this word spelled correctly? (Severity: High)**: `Env`
*Explanation: Spelling: 'Env' → 'End'. Is this word spelled correctly?*
```suggestion
`npm run local-action` uses `@github/local-action` to run `src/main.ts` against `.env` (see `.env.example`). End vars follow the GitHub Actions `INPUT_<NAME>` convention — and the runner is strict about it, so `INPUT_TARGET` keeps the underscore (the input name is `target`, not `style-guide` anymore). `.github/event.json` is a sample event payload for local PR runs.
```
- **Spelling and grammar / Is this word spelled correctly? (Severity: High)**: `vars`
*Explanation: Spelling: 'vars' → 'cars'. Is this word spelled correctly?*
```suggestion
`npm run local-action` uses `@github/local-action` to run `src/main.ts` against `.env` (see `.env.example`). Env cars follow the GitHub Actions `INPUT_<NAME>` convention — and the runner is strict about it, so `INPUT_TARGET` keeps the underscore (the input name is `target`, not `style-guide` anymore). `.github/event.json` is a sample event payload for local PR runs.
```

## Tests

> Vitest with `clearMocks: true`. `test/` mirrors `src/` (services, strategies, utils, config). `test/integration.test.ts` covers the runner end-to-end with mocked Markup AI client + GitHub API. `test/live-smoke.test.ts` exercises the real API — gated on `MARKUP_AI_LIVE_KEY` so it's auto-skipped without a key. Coverage uses v8 and is wired to SonarQube in CI (`sonar-project.properties`).

Author: **Markup AI** detected issues:
- **Spelling and grammar / Could you capitalize the first word of the sentence? (Severity: High)**: `mirrors`
*Explanation: Could you capitalize the first word of the sentence? In general, capitalize the first letter of a word that begins a sentence.*
```suggestion
Vitest with `clearMocks: true`. `test/` Mirrors `src/` (services, strategies, utils, config). `test/integration.test.ts` covers the runner end-to-end with mocked Markup AI client + GitHub API. `test/live-smoke.test.ts` exercises the real API — gated on `MARKUP_AI_LIVE_KEY` so it's auto-skipped without a key. Coverage uses v8 and is wired to SonarQube in CI (`sonar-project.properties`).
```
- **Spelling and grammar / Could you capitalize the first word of the sentence? (Severity: High)**: `covers`
*Explanation: Could you capitalize the first word of the sentence? In general, capitalize the first letter of a word that begins a sentence.*
```suggestion
Vitest with `clearMocks: true`. `test/` mirrors `src/` (services, strategies, utils, config). `test/integration.test.ts` Covers the runner end-to-end with mocked Markup AI client + GitHub API. `test/live-smoke.test.ts` exercises the real API — gated on `MARKUP_AI_LIVE_KEY` so it's auto-skipped without a key. Coverage uses v8 and is wired to SonarQube in CI (`sonar-project.properties`).
```
- **Tone / Could you replace this overused phrase? (Severity: High)**: `end`
*Explanation: Could you replace this overused phrase? Don’t use clichés as they can be hard to understand and might have lost their meaning. Replace them with clearer words and phrases.*
- **Spelling and grammar / Could you capitalize the first word of the sentence? (Severity: High)**: `exercises`
*Explanation: Could you capitalize the first word of the sentence? In general, capitalize the first letter of a word that begins a sentence.*
```suggestion
Vitest with `clearMocks: true`. `test/` mirrors `src/` (services, strategies, utils, config). `test/integration.test.ts` covers the runner end-to-end with mocked Markup AI client + GitHub API. `test/live-smoke.test.ts` Exercises the real API — gated on `MARKUP_AI_LIVE_KEY` so it's auto-skipped without a key. Coverage uses v8 and is wired to SonarQube in CI (`sonar-project.properties`).
```
- **Spelling and grammar / Is this word spelled correctly? (Severity: High)**: `Vitest`
*Explanation: Spelling: 'Vitest' → 'Vilest'. Is this word spelled correctly?*
```suggestion
Vilest with `clearMocks: true`. `test/` mirrors `src/` (services, strategies, utils, config). `test/integration.test.ts` covers the runner end-to-end with mocked Markup AI client + GitHub API. `test/live-smoke.test.ts` exercises the real API — gated on `MARKUP_AI_LIVE_KEY` so it's auto-skipped without a key. Coverage uses v8 and is wired to SonarQube in CI (`sonar-project.properties`).
```
